File: src/main/java/org/vanautrui/languages/parsing/astnodes/nonterminal/TermNode.java
package org.vanautrui.languages.parsing.astnodes.nonterminal;

import org.vanautrui.languages.lexing.utils.TokenList;
import org.vanautrui.languages.lexing.tokens.SymbolToken;
import org.vanautrui.languages.parsing.IASTNode;
import org.vanautrui.languages.parsing.astnodes.ITermNode;
import org.vanautrui.languages.parsing.astnodes.IExpressionComputable;
import org.vanautrui.languages.parsing.astnodes.nonterminal.statements.MethodCallNode;
import org.vanautrui.languages.parsing.astnodes.terminal.*;

public class TermNode implements IASTNode, IExpressionComputable {

    //can be one of these:

    //(keyword constant means true|false|null|this

    /*

    integer constant
    string constant
    keyword constant
    '(' expression ')'
    unaryOp term

     */
    public ITermNode termNode;

    //TODO: insert the oher alternatives

    public TermNode(TokenList tokens)throws Exception{

        TokenList copy = new TokenList(tokens);
		try{
			this.termNode=new FloatConstNode(copy);
		}catch(Exception e0){
		    try{
		        this.termNode=new IntConstNode(copy);
		    }catch (Exception e1){
		        try {
		            this.termNode=new StringConstNode(copy);
		        }catch (Exception e2){
		            try {
		                TokenList copy2=new TokenList(copy);

		                copy2.expectAndConsumeOtherWiseThrowException(new SymbolToken("("));
		                this.termNode=new ExpressionNode(copy2);
		                copy2.expectAndConsumeOtherWiseThrowException(new SymbolToken(")"));

		                copy.set(copy2);
		            }catch (Exception e3){
		                try {
		                    this.termNode=new MethodCallNode(copy);
		                }catch (Exception e4){
							try{
								this.termNode=new BoolConstNode(copy);
							}catch (Exception e5){
							    try {
                                    this.termNode = new VariableNode(copy);
                                }catch (Exception e6){
							    	try {
										this.termNode = new ArrayConstantNode(copy);
									}catch (Exception e7){
							    		this.termNode = new CharConstNode(copy);
									}
                                }
							}
							
		                }
		            }
		        }
		    }
		}

        tokens.set(copy);
    }


    @Override
    public String toSourceCode() {
        return termNode.toSourceCode();
    }
}

File: dragon-parser/src/main/parsing/TermNode.cpp
#include <iostream>
#include <vector>
#include <string>
#include <variant>

#include "../commandline/TokenList.hpp"
#include "../commandline/Token.hpp"
#include "../commandline/TokenKeys.hpp"
#include "TermNode.hpp"
#include "CharConstNode.hpp"
#include "IntConstNode.hpp"
#include "BoolConstNode.hpp"
#include "ExpressionNode.hpp"
#include "VariableNode.hpp"
#include "statements/MethodCallNode.hpp"

using namespace std;

TermNode::TermNode(ExpressionNode* expr){
	this->m5 = expr;
}

TermNode::TermNode(TokenList tokens, bool debug) {

	if(debug){
		cout << "TermNode(...)" << endl;
	}

	TokenList copy = TokenList(tokens);

	try {
		//TODO: re-insert later on
		//this->termNode = FloatConstNode(copy);
		throw "later";
	} catch (string e0) {
		try {
			this->m2 = new IntConstNode(copy,debug);
		} catch (string e1) {
			try {
				//a string constant is syntatic sugar.
				//in the parsing stage it is converted to an array of char constants
				//inline the stringConstant and its syntatic sugar
				Token token = tokens.get(0);
				//TODO: re-enable this later on
				/*
				if (token.kind == STRINGCONSTANT) {

					string content = token.getContents();
					TokenList tks = TokenList(vector(), tokens.relPath);
					tks.push_back(BaseToken(LBRACKET));
					for (int i = 0; i < content.size(); i++) {
						tks.push_back(BaseToken(CHARCONSTANT,content.at(i)));
						if (i < content.size() - 1) {
							tks.push_back(BaseToken(COMMA));
						}
					}
					tks.push_back(BaseToken(RBRACKET));
					this->termNode = ArrayConstantNode(tks);

					copy.consume(1);
				} else {
				*/
					throw "could not read stringConstant syntatic sugar";
				//}
			} catch (string e2) {
				try {
					TokenList copy2 = TokenList(copy);

					copy2.expectAndConsumeOtherWiseThrowException(Token(LPARENS));
					this->m5 = new ExpressionNode(copy2,debug);
					copy2.expectAndConsumeOtherWiseThrowException(Token(RPARENS));

					copy.set(copy2);
				} catch (string e3) {
					try {
						this->m4 = new MethodCallNode(copy,debug);
					} catch (string e4) {
						try {
							this->m1 = new BoolConstNode(copy,debug);
						} catch (string e5) {
							try {
								this->m6 = new VariableNode(copy,debug);
							} catch (string e6) {
								//TODO: re-enable later
								/*
								try {
									this->termNode = ArrayConstantNode(copy);
								} catch (string e7) {
									*/
									this->m3 = new CharConstNode(copy,debug);
								//}
							}
						}
					}
				}
			}
		}
	}
	tokens.set(copy);
}

